includes/api/ApiQueryAllUsers.php

```diff
@@ -0,0 +1,383 @@
+			) ) );
+
+			$this->addFields( 'COUNT(*) AS recentedits' );
+
+			$this->addWhere( "rc_log_type IS NULL OR rc_log_type != 'newusers'" );
+			$timestamp = $db->timestamp( wfTimestamp( TS_UNIX ) - $wgActiveUserDays*24*3600 );
+			$this->addWhere( "rc_timestamp >= {$db->addQuotes( $timestamp )}" );
+
+			$this->addOption( 'GROUP BY', $userFieldToSort );
+		}
+
+		$this->addOption( 'LIMIT', $sqlLimit );
+
+		$this->addFields( array(
+			'user_name',
+			'user_id'
+		) );
+		$this->addFieldsIf( 'user_editcount', $fld_editcount );
+		$this->addFieldsIf( 'user_registration', $fld_registration );
+
+		if ( $useIndex ) {
+			$this->addOption( 'USE INDEX', array( 'user' => 'user_name' ) );
+		}
+
+		$res = $this->select( __METHOD__ );
+
+		$count = 0;
+		$lastUserData = false;
+		$lastUser = false;
+		$result = $this->getResult();
+
+		//
+		// This loop keeps track of the last entry.
+		// For each new row, if the new row is for different user then the last, the last entry is added to results.
+		// Otherwise, the group of the new row is appended to the last entry.
+		// The setContinue... is more complex because of this, and takes into account the higher sql limit
+		// to make sure all rows that belong to the same user are received.
+
+		foreach ( $res as $row ) {
+			$count++;
+
+			if ( $lastUser !== $row->user_name ) {
+				// Save the last pass's user data
+				if ( is_array( $lastUserData ) ) {
+					$fit = $result->addValue( array( 'query', $this->getModuleName() ),
+							null, $lastUserData );
+
+					$lastUserData = null;
+
+					if ( !$fit ) {
+						$this->setContinueEnumParameter( 'from',
+								$this->keyToTitle( $lastUserData['name'] ) );
+						break;
+					}
+				}
+
+				if ( $count > $limit ) {
+					// We've reached the one extra which shows that there are additional pages to be had. Stop here...
+					$this->setContinueEnumParameter( 'from', $this->keyToTitle( $row->user_name ) );
+					break;
+				}
+
+				// Record new user's data
+				$lastUser = $row->user_name;
+				$lastUserData = array(
+					'userid' => $row->user_id,
+					'name' => $lastUser,
+				);
+				if ( $fld_blockinfo && !is_null( $row->ipb_by_text ) ) {
+					$lastUserData['blockedby'] = $row->ipb_by_text;
+					$lastUserData['blockreason'] = $row->ipb_reason;
+					$lastUserData['blockexpiry'] = $row->ipb_expiry;
+				}
+				if ( $row->ipb_deleted ) {
+					$lastUserData['hidden'] = '';
+				}
+				if ( $fld_editcount ) {
+					$lastUserData['editcount'] = intval( $row->user_editcount );
+				}
+				if ( $params['activeusers'] ) {
+					$lastUserData['recenteditcount'] = intval( $row->recentedits );
+				}
+				if ( $fld_registration ) {
+					$lastUserData['registration'] = $row->user_registration ?
+						wfTimestamp( TS_ISO_8601, $row->user_registration ) : '';
+				}
+			}
+
+			if ( $sqlLimit == $count ) {
+				// BUG!  database contains group name that User::getAllGroups() does not return
+				// TODO: should handle this more gracefully
```

Comment:
![INFO](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-info.png) Complete the task associated to this "TODO" comment. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://10.8.74.32:9000/coding_rules#rule_key=php%3AS1135)